### ISSUE TITLE
First @overload is picked rather than the correct one

### ISSUE THREAD

**Issue by lexicalunit** [microsoft/pyright#7841]
**Describe the bug**
This is the smallest representative example code that I can create which still shows the bug. If you plop this file into my repository, you should see the bug. The rest of my code can be found here: https://github.com/lexicalunit/spellbot

```python
from __future__ import annotations

from typing import TYPE_CHECKING, overload
from unittest.mock import MagicMock

import discord
import pytest_asyncio
from spellbot.models import Guild, User

if TYPE_CHECKING:
    from tests.fixtures import Factories


def mock_discord_guild(guild: Guild) -> discord.Guild:
    return MagicMock(spec=discord.Guild)


def mock_discord_user(user: User) -> discord.User:
    return MagicMock(spec=discord.User)


@overload
def mock_discord_object(obj: User) -> discord.User: ...


@overload
def mock_discord_object(obj: Guild) -> discord.Guild: ...


def mock_discord_object(obj: User | Guild) -> discord.User | discord.Guild:
    if isinstance(obj, User):
        return mock_discord_user(obj)
    if isinstance(obj, Guild):
        return mock_discord_guild(obj)
    raise NotImplementedError


@pytest_asyncio.fixture
async def guild(factories: Factories) -> Guild:
    return factories.guild.create()


@pytest_asyncio.fixture
async def discord_guild(guild: Guild) -> discord.Guild:
    return mock_discord_object(guild)  # <-- BUG IS HERE: The incorrect override is selected
```

I tried to factor out my code as much as possible, but the following code does **NOT** have the bug:

```python
from __future__ import annotations

from typing import TYPE_CHECKING, overload
from unittest.mock import MagicMock

import discord
import pytest_asyncio


class Base: ...


class Guild(Base): ...


class User(Base): ...


if TYPE_CHECKING:
    from tests.fixtures import Factories


def mock_discord_guild(guild: Guild) -> discord.Guild:
    return MagicMock(spec=discord.Guild)


def mock_discord_user(user: User) -> discord.User:
    return MagicMock(spec=discord.User)


@overload
def mock_discord_object(obj: User) -> discord.User: ...


@overload
def mock_discord_object(obj: Guild) -> discord.Guild: ...


def mock_discord_object(obj: User | Guild) -> discord.User | discord.Guild:
    if isinstance(obj, User):
        return mock_discord_user(obj)
    if isinstance(obj, Guild):
        return mock_discord_guild(obj)
    raise NotImplementedError


@pytest_asyncio.fixture
async def guild(factories: Factories) -> Guild:
    return factories.guild.create()


@pytest_asyncio.fixture
async def discord_guild(guild: Guild) -> discord.Guild:
    return mock_discord_object(guild)
```

So there's something about my `User` and `Guild` objects? 🤔 

There's nothing weird about them though AFAICT. They're pretty simple `SQLAlchemy` DB models. For example:

```python
class Guild(Base):
    """Represents a Discord guild."""

    __tablename__ = "guilds"

    xid: int = cast(int, Column(BigInteger, primary_key=True, nullable=False))
    created_at = Column(
        DateTime,
        nullable=False,
        default=datetime.utcnow,
        server_default=now,
        doc="UTC timestamp when this guild was first created",
    )
 
    [...]
```

Where `Base` is my `declarative_base()`: https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/basic_use.html

**Code or Screenshots**
But, as you can see here, `pyright` is assuming that the call to `mock_discord_object()` is the first overload defined. _If I swap the order of the overloads, then I don't get an error here. But then I would get an error for a similarly defined `user()` function._
![Screenshot 2024-05-03 at 12 29 39 PM](https://github.com/microsoft/pyright/assets/1903876/eeefc4de-f59e-4ada-badd-f3bc9869d817)

```sh
➜ pyright tests/actions/test_foo.py
/Users/me/spellbot/tests/actions/test_foo.py
  /Users/me/spellbot/tests/actions/test_foo.py:45:12 - error: Expression of type "User" is incompatible with return type "Guild"
    "User" is incompatible with "Guild" (reportReturnType)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
Both.

- Pylance v2024.4.106 (pre-release)
- pyright 1.1.361

`pyproject.toml`:
```toml
[tool.pyright]
enableTypeIgnoreComments = true
reportMissingParameterType = true
reportMissingTypeArgument = true
reportUnnecessaryTypeIgnoreComment = "error"
typeCheckingMode = "basic"
```

**Comment by erictraut** [comment]
Since you were unable to come up with a self-contained example, it will take me a while to investigate this. This is unlikely to be a bug in pyright.

My best guess is that one of the classes involved in your code derives from `Any`. In such a case, a type checker cannot determine type compatibility between the class and others, so it generally needs to assume compatibility with all classes. Please take a look at the classes involved in your example to see if any of them use a dynamic or unknown base class.

**Comment by lexicalunit** [comment]
> My best guess is that one of the classes involved in your code derives from `Any`. In such a case, a type checker cannot determine type compatibility between the class and others, so it generally needs to assume compatibility with all classes. Please take a look at the classes involved in your example to see if any of them use a dynamic or unknown base class.

```python
def declarative_base(
    *,
    metadata: Optional[MetaData] = None,
    mapper: Optional[Callable[..., Mapper[Any]]] = None,
    cls: Type[Any] = object,
    name: str = "Base",
    class_registry: Optional[clsregistry._ClsRegistryType] = None,
    type_annotation_map: Optional[_TypeAnnotationMapType] = None,
    constructor: Callable[..., None] = _declarative_constructor,
    metaclass: Type[Any] = DeclarativeMeta,
) -> Any:
```

From: https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/decl_api.py#L991

😭
